vernac/lemmas.ml

```diff
@@ -178,56 +178,121 @@ let start_lemma_with_initialization ?hook ~poly ~scope ~kind ~udecl sigma recgua
-    ()
+(* Support for mutually proved theorems *)
+
+(* XXX: Most of this does belong to Declare, due to proof_entry manip *)
+module MutualEntry : sig
+
+  (* We keep this type abstract and to avoid uncontrolled hacks *)
+  type t
+
+  val variable : other_thms:Recthm.t list -> Entries.parameter_entry -> t
+
+  val adjust_guardness_conditions
+    : other_thms:Recthm.t list
+    -> possible_indexes:int list list
+    -> Evd.side_effects Declare.proof_entry
+    -> t
+
+  val declare_mutind
+    (* Common to all recthms *)
+    : ?fix_exn:(Exninfo.iexn -> Exninfo.iexn)
+    -> scope:DeclareDef.locality
+    -> poly:bool
+    -> kind:Decls.logical_kind
+    -> hook:DeclareDef.Hook.t option
+    -> uctx:UState.t
+    -> ?hook_data:DeclareDef.Hook.t * UState.t * (Names.Id.t * Constr.t) list
+    -> udecl:UState.universe_decl
+    (* Only for the first constant, introduced by compat *)
+    -> ubind:UnivNames.universe_binders
+    -> name:Id.t
+    -> impargs:Impargs.manual_implicits
+    -> t
+    -> Names.GlobRef.t list
+
+end = struct
+
+  (* Body with the fix *)
+  type et =
+    | NoBody of Entries.parameter_entry
+    | Single of Evd.side_effects Declare.proof_entry
+    | Mutual of Evd.side_effects Declare.proof_entry
+
+  type t =
+    { other_thms : Recthm.t list
+    ; entry : et
+    }
+
+  let variable ~other_thms t = { other_thms; entry = NoBody t }
+
+  (* XXX: Refactor this with the code in
+     [ComFixpoint.declare_fixpoint_generic] *)
+  let guess_decreasing env possible_indexes ((body, ctx), eff) =
+    let open Constr in
+    match Constr.kind body with
+    | Fix ((nv,0),(_,_,fixdefs as fixdecls)) ->
+      let env = Safe_typing.push_private_constants env eff.Evd.seff_private in
+      let indexes = Pretyping.search_guard env possible_indexes fixdecls in
+      (mkFix ((indexes,0),fixdecls), ctx), eff
+    | _ -> (body, ctx), eff
+
+  let adjust_guardness_conditions ~other_thms ~possible_indexes const =
+    let entry = match possible_indexes with
+    | [] ->
+      (* Not a recursive statement *)
+      Single const
+    | possible_indexes ->
+      (* Try all combinations... not optimal *)
+      let env = Global.env() in
+      let pe = Declare.Internal.map_entry_body const
+          ~f:(guess_decreasing env possible_indexes)
+      in
+      Mutual pe
+    in { other_thms; entry }
+
+  let rec select_body i t =
+    let open Constr in
+    match Constr.kind t with
+    | Fix ((nv,0),decls) -> mkFix ((nv,i),decls)
+    | CoFix (0,decls) -> mkCoFix (i,decls)
+    | LetIn(na,t1,ty,t2) -> mkLetIn (na,t1,ty, select_body i t2)
+    | Lambda(na,ty,t) -> mkLambda(na,ty, select_body i t)
+    | App (t, args) -> mkApp (select_body i t, args)
+    | _ ->
+      CErrors.anomaly
+        Pp.(str "Not a proof by induction: " ++
+            Termops.Internal.debug_print_constr (EConstr.of_constr t) ++ str ".")
+
+  let declare_mutind ?fix_exn ~scope ~poly ~kind ~hook ~uctx ?hook_data ~udecl ~ubind ?typ ~name ~impargs mutpe i =
```

Comment:
Should we assert i=0 in the non Mutual cases?

Reply:
Sure why not, in principle nothing limits this code to be used in mutual, but not recursive defs tho.

Reply:
I've added it but unsure this will remain once all code is unified [in depending PR]

Reply:
This is actually incorrect for mutual coinductives [which are handled this way]

Reply:
Indeed it is not clear we are doing the right thing for cofixpoints; CI passes tho.

Reply:
Seems we are doing the right thing, AFAICT.